tests/ballerina-test/src/test/java/org/ballerinalang/test/types/xml/XMLSecurityTest.java

```diff
@@ -0,0 +1,58 @@
+/*
+ *  Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  WSO2 Inc. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.ballerinalang.test.types.xml;
+
+import org.ballerinalang.model.util.XMLUtils;
+import org.ballerinalang.model.values.BXML;
+import org.testng.Assert;
+import org.testng.annotations.Test;
+
+/**
+ * Test cases for XML security.
+ */
+public class XMLSecurityTest {
+
+    @Test
+    public void testExternalEntityInjection() {
+        String xmlString = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>" +
+                "<!DOCTYPE foo [" +
+                "<!ELEMENT foo ANY >" +
+                "<!ENTITY xxe SYSTEM \"https://www.w3schools.com/xml/note.xml\" >]>" +
+                "<foo>&xxe;</foo>";
+        BXML xmlDocument = XMLUtils.parse(xmlString);
+        Assert.assertEquals(xmlDocument.toString(), "<foo></foo>");
+    }
+
+    @Test (timeOut = 10000, expectedExceptions = org.ballerinalang.util.exceptions.BallerinaException.class)
+    public void testEntityExpansion() {
+        String xmlString = "<?xml version=\"1.0\"?>\n" +
+                "<!DOCTYPE lolz [\n" +
+                "<!ENTITY lol \"lol\">\n" +
+                "<!ENTITY lol2 \"&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;&lol;\">\n" +
+                "<!ENTITY lol3 \"&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;&lol2;\">\n" +
+                "<!ENTITY lol4 \"&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;&lol3;\">\n" +
+                "<!ENTITY lol5 \"&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;&lol4;\">\n" +
+                "<!ENTITY lol6 \"&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;&lol5;\">\n" +
+                "<!ENTITY lol7 \"&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;&lol6;\">\n" +
+                "<!ENTITY lol8 \"&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;&lol7;\">\n" +
+                "<!ENTITY lol9 \"&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;&lol8;\">\n" +
+                "]>\n" +
+                "<lolz>&lol9;</lolz>";
+        XMLUtils.parse(xmlString).toString();
```

Comment:
Do we need toString() here?

Reply:
Yes. Otherwise the XML parser (axiom) will try to be smart and will not parser the XML string, since we do not use it anywhere. By doing toString() we are avoiding that since the parser will have to actually parse the original XML to to the toString() operations. This is based on testing I did.

Reply:
Done.

Reply:
Does it mean, is the real case, if someone did not call toString, the attack will not happen?

Reply:
If someone does not use the output of "XMLUtils.parse(xmlString)" to do any operations, the attack will not happen (and actual parsing of XML string will not happen). This seems to be a optimization done in Axiom to avoid unnecessarily parsing strings to XML if no one is going to use the result afterwards.